psi4/src/psi4/libscf_solver/sad.cc

```diff
@@ -167,35 +167,42 @@ SharedMatrix SADGuess::form_D_AO() {
 
     // Spin occupations per atom, to be determined by Hund's Rules
     // or user input
-    std::vector<int> nalpha(molecule_->natom(), 0);
-    std::vector<int> nbeta(molecule_->natom(), 0);
+    std::vector<double> nalpha(molecule_->natom(), 0);
+    std::vector<double> nbeta(molecule_->natom(), 0);
     std::vector<int> nelec(molecule_->natom(), 0);
-    std::vector<int> nhigh(molecule_->natom(), 0);
-    int tot_elec = 0;
 
     // Ground state high spin occupency array, atoms 0 to 36 (see Giffith's Quantum Mechanics, pp. 217)
     // For 37 to 86, save for f-block: Atomic, Molecular, & Optical Physics Handbook, Ed. Gordon W. F. Drake, American
     // Institute of Physics, Woodbury, New York, USA, 1996.
-    const int reference_S[] = {0, 1, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 3, 6, 5, 4, 3, 2,
-                               1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 5, 6, 5, 4, 3, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1,
-                               0, 3, 4, 5, 6, 7, 8, 5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0};
-    const int MAX_Z = 86;
+    const std::vector<int> reference_S = {0, 1, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1,
+                                          2, 3, 6, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 2, 5, 6, 5,
+                                          4, 3, 0, 1, 0, 1, 2, 3, 2, 1, 0, 1, 0, 1, 0, 3, 4, 5, 6, 7, 8, 5,
+                                          4, 3, 2, 1, 0, 1, 2, 3, 4, 5, 4, 3, 2, 1, 0, 1, 2, 3, 2, 1, 0};
 
     if (print_ > 1) outfile->Printf("  Determining Atomic Occupations\n");
 
     for (int A = 0; A < molecule_->natom(); A++) {
         int Z = molecule_->Z(A);
-        if (Z > MAX_Z) {
-            throw std::domain_error(" Only Atoms up to 86 (Rn) are currently supported with SAD Guess");
-        }
-        nhigh[A] = reference_S[Z];
+        // Assuming neutral atoms for now
         nelec[A] = Z;
-        tot_elec += nelec[A];
-        nbeta[A] = (nelec[A] - nhigh[A]) / 2;
-        nalpha[A] = nelec[A] - nbeta[A];
-        if (print_ > 1)
-            outfile->Printf("  Atom %d, Z = %d, nelec = %d, nhigh = %d, nalpha = %d, nbeta = %d\n", A, Z, nelec[A],
-                            nhigh[A], nalpha[A], nbeta[A]);
+
+        if (options_.get_bool("SAD_SPIN_AVERAGE")) {
+            // Spin-averaged occupations
+            nalpha[A] = nbeta[A] = 0.5 * nelec[A];
+        } else {
+            // Target ground spin state
+            if (Z >= reference_S.size()) {
+                std::ostringstream err;
+                err << " Only atoms up to Z = " << reference_S.size() - 1
```

Comment:
I don't think the `-1` is needed here, because of the extra element at the front

Reply:
It is. The first element is padding for ```Z=0``` which doesn't exist. The spin state for ```Z``` is ```reference_S[Z]```. The last valid ```Z``` is the one which gives ```reference_S[reference_S.size()-1]``` i.e. ```Z=reference_S.size()-1```.